servant-purescript/src/Servant/PureScript.hs

```diff
@@ -0,0 +1,229 @@
+        , "import Network.HTTP.Affjax"
+        , "import Network.HTTP.Method"
+        , "import Network.HTTP.RequestHeader"
+        , "import " <> (settings ^. utilModuleName) <> " (encodeURIComponent)"
+        , ""
+        , T.intercalate "\n" (generatePS settings <$> reqs)
+        ]
+
+-- | Generate a single PureScript function for an AJAX request.
+-- To prevent conflicts, generates a unique function name for every available
+-- function name and set of captures.
+generatePS
+    :: PSSettings -- ^ PureScript rendering settings
+    -> F.Req -- ^ AJAX request to render
+    -> ST -- ^ Rendered PureScript
+generatePS settings req = ajaxRequest
+  where
+    args :: [ST]
+    args = captures <> queryArgs <> body <> fmap (fst . snd) headerArgs
+
+    captures :: [ST]
+    captures = fmap F.captureArg . filter F.isCapture $ req ^. F.reqUrl . F.path
+
+    queryArgs :: [ST]
+    queryArgs  = fmap ((<>) "query" . view F.argName) queryParams
+
+    headerArgs :: [(T.Text, (T.Text, F.HeaderArg))]
+    headerArgs = fmap ( decapitalise . F.headerArgName &&&
+                        JS.toValidFunctionName . (<>) "header" . F.headerArgName &&&
+                        id )
+                      (req ^. F.reqHeaders)
+
+    fname :: ST
+    fname = F.camelCase (req ^. F.funcName)
+         <> if null captures then "" else "With"
+         <> T.intercalate "And" (fmap capitalise captures)
+
+    queryParams :: [F.QueryArg]
+    queryParams = req ^.. F.reqUrl . F.queryStr . traverse
+
+    body :: [ST]
+    body = ["body" | req ^. F.reqBody]
+
+    wrapHeaders :: [(T.Text, T.Text)] -> [(T.Text, (T.Text, F.HeaderArg))] -> ST
+    wrapHeaders ihs hs =
+        "[" <>
+        T.intercalate ", " (concat $
+            [wrapImplicitHeader <$> ihs | not $ null ihs] ++
+            [wrapHeader <$> hs | not $ null hs]) <>
+        "]"
+
+    wrapHeader :: (T.Text, (T.Text, F.HeaderArg)) -> ST
+    wrapHeader (h, (_, o)) = "RequestHeader " <> cs (show h) <> " " <> psHeaderArg o
+
+    wrapImplicitHeader :: (T.Text, T.Text) -> ST
+    wrapImplicitHeader (k, v) = "RequestHeader " <> cs (show k) <> " " <> cs (show v)
+
+    implicitHeaderArgs :: [(T.Text, T.Text)]
+    implicitHeaderArgs =
+        [ ("content-type", "application/json")
+        , ("accept", "application/json")
+        ]
+
+    ajaxRequest :: ST
+    ajaxRequest = T.unlines $
+        typeSig :
+        (fname <> argString <> " = affjax $ defaultRequest") :
+        ("    { method = " <> req ^. F.reqMethod) :
+        ("    , url = " <> urlString) :
+        ("    , headers = " <> wrapHeaders implicitHeaderArgs headerArgs) :
+        ["    , content = Just body" | req ^. F.reqBody] ++
+        ["    }"]
+      where
+        typeSig :: ST
+        typeSig = T.concat
+            [ fname
+            , " :: forall eff. "
+            , T.intercalate " -> " typedArgs
+            , if null args then "" else " -> "
+            , "Affjax eff Foreign"
+            ]
+
+        typedArgs :: [ST]
```

Comment:
What's the right way of dealing with types here? Could [this](https://github.com/philopon/haskell-purescript-binding) be it? 

We haven't really figured out the best way of doing codegen for typed languages in servant-foreign yet, and this may be a good opportunity, though purescript is a bit of a special example.